src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final float INCREASE_FACTOR = 1.5f;
+
+    private int size;
+    private int capacity = DEFAULT_CAPACITY;
+    private Object[] list;
+
+    public ArrayList() {
+        this.list = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndexForAdding(index);
+        if (size >= capacity) {
+            increaseCapacity();
+        }
+        if (index < size) {
+            shiftAllLeftFromIndex(index);
+        }
+        list[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) list[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        list[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedEl = (T) list[index];
+        shiftAllRightFromIndex(index);
+        size--;
+        return removedEl;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedEl;
+        for (int i = 0; i < size; i++) {
+            if (element == null && list[i] == null
+                    || list[i] != null && list[i].equals(element)) {
+                removedEl = (T) list[i];
+                remove(i);
+                return removedEl;
+            }
+        }
+        throw new NoSuchElementException("No such element " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void increaseCapacity() {
+        Object[] tempList = list;
+
+        capacity *= INCREASE_FACTOR;
+        list = new Object[capacity];
+
+        for (int i = 0; i < size; i++) {
+            list[i] = tempList[i];
+        }
+    }
+
+    private void validateIndex(int index) {
+        validateIndexToBound(index, size);
+    }
+
+    private void validateIndexForAdding(int index) {
+        validateIndexToBound(index, size + 1);
+    }
+
+    private void validateIndexToBound(int index, int bound) {
+        if (index < 0 || index >= bound) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " out of bounds for length " + size
+            );
```

Comment:
I am not sure of the fact, but it seems that the size - 1 should be used, because u haven't any element in index[size]? Plz, report if I lost my mind.

Reply:
`size` corresponds to the number of elements that we have in list. So, in other words this message says "Hey, you have ten elements, so you can't use index 10, because it is out of bounds."